api/grpcserver/mesh_service.go

```diff
@@ -0,0 +1,360 @@
+	return &pb.CurrentEpochResponse{Epochnum: &pb.SimpleInt{
+		Value: uint64(curLayer.GetEpoch(uint16(s.LayersPerEpoch))),
+	}}, nil
+}
+
+// NetID returns the network ID
+func (s MeshService) NetID(ctx context.Context, in *pb.NetIDRequest) (*pb.NetIDResponse, error) {
+	log.Info("GRPC MeshService.NetId")
+	return &pb.NetIDResponse{Netid: &pb.SimpleInt{
+		Value: uint64(s.NetworkID),
+	}}, nil
+}
+
+// EpochNumLayers returns the number of layers per epoch (a network parameter)
+func (s MeshService) EpochNumLayers(ctx context.Context, in *pb.EpochNumLayersRequest) (*pb.EpochNumLayersResponse, error) {
+	log.Info("GRPC MeshService.EpochNumLayers")
+	return &pb.EpochNumLayersResponse{Numlayers: &pb.SimpleInt{
+		Value: uint64(s.LayersPerEpoch),
+	}}, nil
+}
+
+// LayerDuration returns the layer duration in seconds (a network parameter)
+func (s MeshService) LayerDuration(ctx context.Context, in *pb.LayerDurationRequest) (*pb.LayerDurationResponse, error) {
+	log.Info("GRPC MeshService.LayerDuration")
+	return &pb.LayerDurationResponse{Duration: &pb.SimpleInt{
+		Value: uint64(s.LayerDurationSec),
+	}}, nil
+}
+
+// MaxTransactionsPerSecond returns the max number of tx per sec (a network parameter)
+func (s MeshService) MaxTransactionsPerSecond(ctx context.Context, in *pb.MaxTransactionsPerSecondRequest) (*pb.MaxTransactionsPerSecondResponse, error) {
+	log.Info("GRPC MeshService.MaxTransactionsPerSecond")
+	return &pb.MaxTransactionsPerSecondResponse{Maxtxpersecond: &pb.SimpleInt{
+		Value: uint64(s.TxsPerBlock * s.LayerAvgSize / s.LayerDurationSec),
+	}}, nil
+}
+
+// QUERIES
+
+// AccountMeshDataQuery returns account data
+func (s MeshService) AccountMeshDataQuery(ctx context.Context, in *pb.AccountMeshDataQueryRequest) (*pb.AccountMeshDataQueryResponse, error) {
+	log.Info("GRPC MeshService.AccountMeshDataQuery")
+
+	// This is the latest layer we're aware of
+	//latestLayer := s.Mesh.LatestLayer()
+	startLayer := types.LayerID(in.MinLayer)
+
+	if startLayer > s.Mesh.LatestLayer() {
+		return nil, status.Errorf(codes.InvalidArgument, "`LatestLayer` must be less than or equal to latest layer")
+	}
+	if in.Filter == nil {
+		return nil, status.Errorf(codes.InvalidArgument, "`Filter` must be provided")
+	}
+	if in.Filter.AccountId == nil {
+		return nil, status.Errorf(codes.InvalidArgument, "`Filter.AccountId` must be provided")
+	}
+
+	// Gather transaction data
+	//txs := &pb.AccountMeshData{DataItem: &pb.AccountMeshData_Transaction{Transaction: }
+	addr := types.BytesToAddress(in.Filter.AccountId.Address)
+	res := &pb.AccountMeshDataQueryResponse{}
+	if in.Filter.AccountMeshDataFlags&uint32(pb.AccountMeshDataFlag_ACCOUNT_MESH_DATA_FLAG_TRANSACTIONS) != 0 {
+		meshTxIds := s.getTxIdsFromMesh(startLayer, addr)
+		mempoolTxIds := s.Mempool.GetTxIdsByAddress(addr)
+
+		// Look up full data for all unique txids
+		txs, missing := s.Mesh.GetTransactions(append(meshTxIds, mempoolTxIds...))
+
+		// TODO: Do we ever expect txs to be missing here?
```

Comment:
@antonlerner is it an error if there are missing transactions here? How should we handle this?

Reply:
I'd return error with missing ids, to my understanding this can only happen if a client queries for wrong account or has messed up the addresses in its request. this should return an error to the client even if some tx are found since it probably indicates that the query was built in a wrong way

Reply:
It doesn't seem like it could be a query issue. The query only includes a layerId and an address, and we attempt to read all transactions for that address - there should either be some, or none, right? Why would a layer/the mempool give us txids for txs we can't read from the mesh? This seems like an "internal error" to me (something involving database consistency). In any case, we agree that this should be an error.